File: app/django/api/serializers/accounts.py
from rest_framework import serializers

from accounts.models import User, StaffAuth, Profile, Todo


# Accounts --------------------------------------------------------------------------
class StaffAuthInUserSerializer(serializers.ModelSerializer):
    class Meta:
        model = StaffAuth
        fields = ('pk', 'company', 'is_staff', 'assigned_project', 'allowed_projects', 'contract',
                  'payment', 'notice', 'project_cash', 'project_docs', 'project', 'company_cash',
                  'company_docs', 'human_resource', 'company_settings', 'auth_manage')


class ProfileInUserSerializer(serializers.ModelSerializer):
    class Meta:
        model = Profile
        fields = ('pk', 'name', 'birth_date', 'cell_phone', 'image')


class UserSerializer(serializers.ModelSerializer):
    password = serializers.CharField(
        write_only=True,
        required=True,
        help_text='변경할 필요가 없으면 비워 두십시오.',
        style={'input_type': 'password', 'placeholder': '비밀번호'}
    )
    staffauth = StaffAuthInUserSerializer()
    profile = ProfileInUserSerializer(read_only=True)

    class Meta:
        model = User
        fields = ('pk', 'email', 'username', 'is_active', 'is_superuser',
                  'date_joined', 'password', 'staffauth', 'profile')

    def save(self):
        instance = User(email=self.validated_data['email'],
                        username=self.validated_data['username'])
        password = self.validated_data['password']
        instance.set_password(password)
        instance.save()
        self.instance = instance
        return self.instance


class ProfileSerializer(serializers.ModelSerializer):
    image = serializers.ImageField(use_url=True)

    class Meta:
        model = Profile
        fields = ('pk', 'user', 'name', 'birth_date', 'cell_phone', 'image')


class TodoSerializer(serializers.ModelSerializer):
    class Meta:
        model = Todo
        fields = ('pk', 'user', 'title', 'completed', 'created_at', 'updated_at', 'soft_deleted')

File: app/vue/src/store/types/accounts.ts
export declare interface User {
  pk?: number
  email: string
  username: string
  is_active: boolean
  is_superuser: boolean
  date_joined: string
  staffauth: StaffAuth | null
  profile: null | {
    pk: number
    name: string
    birth_date: string
    cell_phone: string
  }
  last_login: string | null
}

type Auth = '0' | '1' | '2'

export interface StaffAuth {
  pk?: number
  user?: number
  company: number | null
  is_staff: boolean
  is_project_staff: boolean
  allowed_projects: number[]
  assigned_project: number | null
  contract: Auth
  payment: Auth
  notice: Auth
  project_cash: Auth
  project_docs: Auth
  project: Auth
  company_cash: Auth
  company_docs: Auth
  human_resource: Auth
  company_settings: Auth
  auth_manage: Auth
}

export type Profile = {
  [key: string]: undefined | number | number[] | null | string | File
  pk?: number | null
  user?: number | null
  name: string
  birth_date: string
  cell_phone: string
  image?: File | string | null
  like_posts?: number[]
  like_comments?: number[]
  blame_posts?: number[]
  blame_comments?: number[]
}

export interface Scrape {
  pk?: number
  user: number
  post: {
    pk: number
    board: number
    board_name: string
    project: number | null
    title: string
  }
  title: string
  created: string
}

export interface Todo {
  pk?: number
  user?: number
  title?: string
  completed?: boolean
  soft_deleted?: boolean
}
